src/core/src/mapping.rs

```diff
@@ -18,163 +18,268 @@
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub enum Error {
+    /// The requested mapping access did not match the expected usage.
+    InvalidAccess(Access, factory::Usage),
+}
+
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub struct RawInner<R: Resources> {
+    pub resource: R::Mapping,
+    pub buffer: handle::RawBuffer<R>,
+    pub access: Access,
+    pub status: Status<R>,
 }
 
-impl<'a, T: Copy, R: Resources, F: Factory<R>> Drop for Readable<'a, T, R, F> where
-    F::Mapper: 'a
-{
+impl<R: Resources> Drop for RawInner<R> {
     fn drop(&mut self) {
-        self.factory.unmap_buffer_raw(self.raw.clone())
+        self.buffer.was_unmapped();
     }
 }
 
-/// A handle to a writable map, which only allows setting elements.
-pub struct Writable<'a, T: Copy, R: 'a + Resources, F: 'a + Factory<R>> where
-    F::Mapper: 'a
-{
-    raw: F::Mapper,
-    len: usize,
-    factory: &'a mut F,
-    phantom_t: PhantomData<T>
-}
+/// Raw mapping providing status tracking
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub struct Raw<R: Resources>(Mutex<RawInner<R>>);
 
-impl<'a, T: Copy, R: Resources, F: Factory<R>> Writable<'a, T, R, F> where
-    F::Mapper: 'a
-{
-    /// Set a value in the buffer
-    pub fn set(&mut self, idx: usize, val: T) {
-        if idx >= self.len {
-            panic!("Tried to write out of bounds to a WritableMapping!")
+#[allow(missing_docs)]
+impl<R: Resources> Raw<R> {
+    pub fn new(res: R::Mapping, access: Access, buf: &handle::RawBuffer<R>) -> Self {
+        Raw(Mutex::new(RawInner {
+            resource: res,
+            buffer: buf.clone(),
+            access: access,
+            status: Status::clean(),
+        }))
+    }
+
+    pub fn access(&self) -> Option<MutexGuard<RawInner<R>>> {
+        self.0.try_lock().ok()
+    }
+
+    unsafe fn read<T: Copy>(&self, len: usize) -> Reader<R, T> {
+        let mut inner = self.access().unwrap();
+        R::Mapping::before_read(&mut inner);
+        inner.status.access();
+
+        Reader {
+            slice: inner.resource.slice(len),
+            inner: inner,
         }
-        unsafe { self.raw.set(idx, val); }
     }
-    /// Returns a mutable slice of the specified length.
-    pub fn to_mut_slice(&mut self) -> &mut [T] {
-        unsafe { self.raw.to_mut_slice(self.len) }
+
+    unsafe fn write<T: Copy>(&self, len: usize) -> Writer<R, T> {
+        let mut inner = self.access().unwrap();
+        R::Mapping::before_write(&mut inner);
+        inner.status.write_access();
+
+        Writer {
+            len: len,
+            inner: inner,
+            phantom: PhantomData,
+        }
     }
-}
 
-impl<'a, T: Copy, R: Resources, F: Factory<R>> Drop for Writable<'a, T, R, F> where
-    F::Mapper: 'a
-{
-    fn drop(&mut self) {
-        self.factory.unmap_buffer_raw(self.raw.clone())
+    unsafe fn read_write<T: Copy>(&self, len: usize) -> RWer<R, T> {
+        let mut inner = self.access().unwrap();
+        R::Mapping::before_read(&mut inner);
```

Comment:
huh, I'm surprised this works. Typically Rust required us to do `<X::Y as Z>::fun`. If that's fixed, we might be able to simplify some more code here and there (outside of this PR, even outside of GFX).